### ISSUE TITLE
Compiling error for `refactoring_sandbox` branch on Windows

### ISSUE THREAD

**Issue by BigRoy** [Autodesk/maya-usd#50]
### Issue

I'm trying to compile the `refactoring_sandbox` branch with the [latest commit](https://github.com/Autodesk/maya-usd/commit/76e3172a276b4494f1f2356d4e00ccf61544c792) however I seem to be unable to compile `maya-usd` on Windows.

I compiled USD with commit `b85ddac2` using:
```
%PYTHON% build_scripts/build_usd.py --build-args boost,"--with-date_time --with-thread --with-system --with-filesystem" --no-maya %USD_BUILD_DIR%
```

Then I tried to compile `maya-usd` yet it failed on some missing `.lib` files, which I [hacked using this solution](https://gist.github.com/BigRoy/4593ef47f39a3589308454088a0a2753#gistcomment-3050688) and then it passed on as it seems it should.

This is what I used to trigger the compile of `maya-usd`:
```
%PYTHON% build.py --maya-location %MAYA_LOCATION% --pxrusd-location %USD_BUILD_DIR% --devkit-location %MAYA_DEVKIT_LOCATION% %USD_MAYA_BUILD_DIR%
```

But then it hit an error linking some code. Specifically it errors on:
```
wrapTranslatorContext.obj : error LNK2005: "class AL::usdmaya::nodes::ProxyShape const volatile * __cdecl boost::get_pointer<class AL::usdmaya::nodes::ProxyShape const volatile >(class AL::usdmaya::nodes::ProxyShape const volatile *)" (??$get_pointer@$$CDVProxyShape@nodes@usdmaya@AL@@@boost@@YAPEDVProxyShape@nodes@usdmaya@AL@@PEDV1234@@Z) already defined in wrapProxyShape.obj [D:\dev\usd\maya-usd-build\build\RelWithDebInfo\plugin\al\lib\AL_USDMaya\_AL_USDMaya.vcxproj]

     Creating library D:/dev/usd/maya-usd-build/build/RelWithDebInfo/plugin/al/lib/AL_USDMaya/RelWithDebInfo/_AL_USDMaya.lib and object D:/dev/usd/maya-usd-build/build/RelWithDebInfo/plugin/al/lib/AL_USDMaya/RelWithDebInfo/_AL_USDMaya.exp

D:\dev\usd\maya-usd-build\build\RelWithDebInfo\plugin\al\lib\AL_USDMaya\RelWithDebInfo\_AL_USDMaya.pyd : fatal error LNK1169: one or more multiply defined symbols found [D:\dev\usd\maya-usd-build\build\RelWithDebInfo\plugin\al\lib\AL_USDMaya\_AL_USDMaya.vcxproj]

     Creating library D:/dev/usd/maya-usd-build/build/RelWithDebInfo/plugin/al/plugin/AL_USDMayaPlugin/RelWithDebInfo/AL_USDMayaPlugin.lib and object D:/dev/usd/maya-usd-build/build/RelWithDebInfo/plugin/al/plugin/AL_USDMayaPlugin/RelWithDebInfo/AL_USDMayaPlugin.exp

  AL_USDMayaPlugin.vcxproj -> D:\dev\usd\maya-usd-build\build\RelWithDebInfo\plugin\al\plugin\AL_USDMayaPlugin\RelWithDebInfo\AL_USDMayaPlugin.mll
```

The full build log is here: [build_log.txt](https://github.com/Autodesk/maya-usd/files/3708689/build_log.txt)

**Comment by kxl-adsk** [comment]
I believe you are compiling with VS2017? We have a fix internally for it, but it's not yet pushed. See the diff below
![image](https://user-images.githubusercontent.com/24443089/66506969-5302c900-ea9c-11e9-9f28-92c554d98917.png)

**Comment by BigRoy** [comment]
Thanks!

> I believe you are compiling with VS2017? 

Yes, I am. :) Running from "Developer Command Prompt for VS 2017"

---

I pushed the text through a Image to Text conversion. This should make it easier for others to try the patch (including the comment for completeness sake + fixed a typo in the comment):
```cpp
// This workaround for a VS compiler bug where we need to explicitly specify
// the conversion to pointer of a class seems no longer needed for VS2017.
// What it means, is that the compiler will properly generate this conversion
// implicitly and cause linker error LNK2005.
//
// In this particular case we got wrapTranslatorContext.obj generating error for
// already defined conversion to a pointer for ProxyShape in wrapProxyShape.obj
//
// The best place to put this fix would be in pxr/base/lib/tf/refPtr.h
// where TF_REFPTR_CONST_VOLATILE_GET is defined, but for now we are
// patching it locally.
#if defined(ARCH_COMPILER_MSVC) && ARCH_COMPILER_MSCV_VERSION <= 1910
    TF_REFPTR_CONST_VOLATILE_GET(ProxyShape)
#endif
```

Anyhow, it didn't seem to resolve my particular issue. I'm seeing a lot of this in my build log too:
```
Unknown compiler version - please run the configure tests and report the results
```

So I assumed that maybe it could be that even with this fix the `ARCH_COMPILER_MSCV_VERSION` is not detected correctly or `ARCH_COMPILER_MSVC` somehow not getting defined. To test that I deleted the full line alltogether so it could never get triggered which according to the fix should be the case for "my case".

Having removed the line completely, suddenly it worked. It seems that my compiler passed the if-statement.

The failed build log seems similar to before: 
[build_log_with_fix.txt](https://github.com/Autodesk/maya-usd/files/3711580/build_log_with_fix.txt)

The one with the full line removed that succeeded continued to the end: 
[build_log_success.txt](https://github.com/Autodesk/maya-usd/files/3711587/build_log_success.txt)

Let me know what input you need from my end to correctly patch it down the line. Or what I'd need to do have it pass like yours.

**Comment by HamedSabri-adsk** [comment]
@BigRoy  The warning seems to be a known issue related to boost library and I agree it is very annoying:

https://developercommunity.visualstudio.com/content/problem/93313/latest-update-breaks-compiling-with-boost.html

https://github.com/Microsoft/CNTK/issues/3344

**Comment by BigRoy** [comment]
Thanks @HamedSabri-adsk - luckily it's just a warning. I'm not totally sure whether that warning relates to the issue of this if statement still passing and triggering the `TF_REFPTR_CONST_VOLATILE_GET` in my building process:
```c++
#if defined(ARCH_COMPILER_MSVC) && ARCH_COMPILER_MSCV_VERSION <= 1910
    TF_REFPTR_CONST_VOLATILE_GET(ProxyShape)
#endif
```

I had to remove the `TF_REFPTR_CONST_VOLATILE_GET` line completely as somehow my build still triggered it with this [fix that was proposed](https://github.com/Autodesk/maya-usd/issues/50#issuecomment-540115672).

**Comment by kxl-adsk** [comment]
The proposed fix works for all our internal builds. Can you tell me what's the version reported by ARCH_COMPILER_MSCV_VERSION and exact VS2017 version?

**Comment by BigRoy** [comment]
I will be back in office on Thursday so I will take a look then. 

What's the easiest way to debug that value, for example to get it into the build logs? Can I make it print it somewhere during compiling?

**Comment by kxl-adsk** [comment]
Add to the same file
```
#define VALUE_TO_STRING(x) #x
#define VALUE(x) VALUE_TO_STRING(x)
#pragma message("ARCH_COMPILER_MSVC_VERSION=" VALUE(ARCH_COMPILER_MSVC_VERSION))
```

**Comment by BigRoy** [comment]
I just pulled the latest `refactoring_sandbox` and added the code you provided to debug. Interestingly enough the build *also* succeeded which the build log confirms with the compiler version:

```
ARCH_COMPILER_MSVC_VERSION=1916
```

As such it must have been some human error when I tested the fix before and it is now fixed in `refactoring_sandbox` branch with this PR https://github.com/Autodesk/maya-usd/pull/58

I'll close this issue as I was able to build the `refactoring_sandbox` branch now without any local changes to the code base. Note that I did build `USD` itself on Windows following [these notes](https://gist.github.com/BigRoy/4593ef47f39a3589308454088a0a2753#gistcomment-3050688).
